mesonbuild/modules/rust.py

```diff
@@ -142,9 +157,28 @@ def test(self, state: ModuleState, args: T.Tuple[str, BuildTarget], kwargs: Func
         new_target_kwargs = base_target.kwargs.copy()
         # Don't mutate the shallow copied list, instead replace it with a new
         # one
-        new_target_kwargs['rust_args'] = new_target_kwargs.get('rust_args', []) + ['--test']
+        new_target_kwargs['rust_args'] = \
+            new_target_kwargs.get('rust_args', []) + kwargs['rust_args'] + ['--test']
         new_target_kwargs['install'] = False
         new_target_kwargs['dependencies'] = new_target_kwargs.get('dependencies', []) + kwargs['dependencies']
+        new_target_kwargs['link_with'] = new_target_kwargs.get('link_with', []) + kwargs['link_with']
+
+        # If any built targets being linked with use C++, we need to add
+        # arguments to link with the C++ stdlib.
+        links_with = itertools.chain(
+            new_target_kwargs['link_with'],
+            itertools.chain.from_iterable(d.get_link_with_recursive() for d in new_target_kwargs['dependencies']))
+        l: LibTypes
+        for l in links_with:
+            if isinstance(l, (CustomTarget, CustomTargetIndex)):
+                mlog.warning('Rust test:', name, 'is linking with CustomTarget library, it cannot be determined if this is a C++ ',
+                             'library. If it is, you may need to link against libc++ or libstdc++ manually.',
+                             once=True)
+                continue
+            if 'cpp' in l.compilers:
+                cpp = state.environment.coredata.compilers.host['cpp']
+                new_target_kwargs['link_args'] = new_target_kwargs.get('link_args', []) + cpp.language_stdlib_only_link_flags(state.environment)
```

Comment:
Feels like this problem is not limited to Rust test linking to a C++ library. This can happen for any executable that links to libraries that require an std. I see `language_stdlib_only_link_flags()` is implemented for fortran too.

I would expect this to be handled already by `get_clink_dynamic_linker_and_stdlibs()`, but maybe its implementation is broken?

Reply:
Also, isn't the problem only when we link a static library that requires libc++? If it was a shared library the linker would find libraries it links to already, no?

Reply:
Rust doesn't use that code, only C-like languages do. Maybe this does need to be handled in the backend though

Reply:
In any case, would like to see a unit test.

Reply:
I am now curious as to how C++ and Fortarn link together...